File: AtCoder/abc/abc192/d.py
def basen(x, d):
    ret = 0
    base = 1
    n = len(x)

    for i in reversed(range(n)):
        ret += int(x[i]) * base
        base *= d

    return ret

x = input()
m = int(input())

d = 0
for c in x:
    d = max(d, int(c))

if(basen(x, d+1) > m):
    print(0)
    exit()

if(len(x) == 1):
    print(1)
    exit()

ok = d+1
ng = 1<<61
while ng-ok > 1:
    mid = (ok + ng) // 2

    if(basen(x, mid) > m): ng = mid
    else: ok = mid

print(ok - d)

File: AtCoder/abc/abc192/d.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;
using ll = long long;
const ll INF = 9223372036854775807ll;

ll basen(string x, ll d){
    ll ret = 0;
    ll base = 1;
    ll n = x.size();

    for(ll i = n-1; i >= 0; --i){
        ret += (x[i]-'0') * base;
        base *= d;
        if(base < 0) return -1;
    }

    return ret;
}


int main(){
    cin.tie(nullptr);
    ios::sync_with_stdio(false);

    string x;
    ll m;
    cin >> x >> m;

    ll d = 0;
    for(auto c: x){
        d = max(d, ll(c-'0'));
    }

    if(basen(x, d+1)==-1 || basen(x, d+1)>m){
        cout << 0 << endl;
        return 0;
    }

    ll ok = d+1;
    ll ng = INF - ok;
    while(ng-ok > 1){
        ll mid = (ok + ng) / 2;

        ll val = basen(x, mid);

        cout << ok << " " << mid << " " << ng << " " << val << endl;
        if(val==-1 || val>m) ng = mid;
        else ok = mid;
    }

    cout << ok - d << endl;

    return 0;
}
